File: bin/contrast-login.py
"""
Pixel side of `bin/contrast-login.mjs`.

Takes the manifest that script writes, averages each background sample and
reports the WCAG 2.1 contrast ratio against the text colour that was measured
on the page.

The averaging matters for these screens: a gradient behind a line of text has
no single value, and the blur behind frosted glass has whatever the field
underneath it contributed. An average is the honest summary of a backdrop that
varies, and the variation across one line of text is small enough that it does
not hide a failure at either end.
"""

import json
import sys

from PIL import Image

AA_NORMAL = 4.5
AA_LARGE = 3.0

# Text that is 24px or heavier-and-18.66px counts as large under 1.4.3. The
# headings and the stage type are both well past it, so they are held to 3:1;
# everything else to 4.5:1.
LARGE = {'heading', 'stage name'}


def channel(value: float) -> float:
    value /= 255

    return value / 12.92 if value <= 0.04045 else ((value + 0.055) / 1.055) ** 2.4


def luminance(rgb: tuple[float, float, float]) -> float:
    r, g, b = (channel(c) for c in rgb)

    return 0.2126 * r + 0.7152 * g + 0.0722 * b


def ratio(a: tuple[float, float, float], b: tuple[float, float, float]) -> float:
    first, second = sorted((luminance(a), luminance(b)), reverse=True)

    return (first + 0.05) / (second + 0.05)


def parse(colour: list[float]) -> tuple[float, float, float]:
    return tuple(colour[:3])


def average(path: str) -> tuple[float, float, float]:
    image = Image.open(path).convert('RGB')
    pixels = list(image.getdata())

    return tuple(sum(c) / len(pixels) for c in zip(*pixels))


def main(manifest: str) -> int:
    samples = json.load(open(manifest))

    rows = []
    failures = 0

    for sample in samples:
        text = parse(sample['colour'])
        background = average(sample['file'])
        measured = ratio(text, background)

        required = AA_LARGE if sample['name'] in LARGE else AA_NORMAL
        passed = measured >= required

        failures += not passed

        rows.append(
            (
                sample['layout'],
                sample['mode'],
                sample['name'],
                f'{measured:.2f}:1',
                f'{required:.1f}:1',
                'pass' if passed else 'FAIL',
            )
        )

    widths = [max(len(str(row[i])) for row in rows) for i in range(6)]

    for row in rows:
        print('  '.join(str(cell).ljust(widths[i]) for i, cell in enumerate(row)))

    print()
    print(f'{len(rows)} pairs measured, {failures} below AA')

    return 1 if failures else 0


if __name__ == '__main__':
    sys.exit(main(sys.argv[1]))

File: bin/contrast-login.mjs
/*
 * Measured contrast for the sign-in compositions.
 *
 * The colour report in `bin/contrast-report.php` works from the palette, which
 * is enough for flat surfaces. These screens are not flat: two of them put
 * text over a gradient and one puts it over frosted glass, and a ratio
 * calculated against a nominal background would not be a measurement of
 * anything.
 *
 * So this reads pixels. For every piece of text on the screen it takes the
 * computed colour, hides the glyphs, photographs the box they occupied, and
 * averages what is behind them. The ratio is then WCAG 2.1 1.4.3 against the
 * real backdrop, including whatever the blur picked up.
 *
 *     php vendor/bin/testbench serve --host=127.0.0.1 --port=8321
 *     node bin/contrast-login.mjs
 */

import { execFileSync } from 'node:child_process'
import { mkdtempSync, writeFileSync } from 'node:fs'
import { tmpdir } from 'node:os'
import { join } from 'node:path'
import { chromium } from 'playwright'

const BASE = process.env.MIA_BASE ?? 'http://127.0.0.1:8321'

const LAYOUTS = ['card', 'split', 'bleed', 'editorial', 'portal']

const TARGETS = [
    ['heading', '.fi-simple-header-heading'],
    ['subheading', '.fi-simple-header-subheading'],
    ['registration link', '.fi-simple-header-subheading a'],
    ['field label', '.fi-fo-field-wrp-label'],
    ['recovery link', '.fi-fo-field-wrp-hint a'],
    ['remember me', '.fi-fo-checkbox .fi-fo-field-wrp-label'],
    ['submit', '.fi-form-actions .fi-btn-label'],
    ['error message', '.fi-fo-field-wrp-error-message'],
    ['stage name', '.fi-mia-login-stage-name'],
    ['stage tagline', '.fi-mia-login-stage-tagline'],
]

const dir = mkdtempSync(join(tmpdir(), 'mia-contrast-'))

const browser = await chromium.launch()
const samples = []

for (const layout of LAYOUTS) {
    execFileSync(
        'php',
        [
            'vendor/bin/testbench',
            'tinker',
            '--execute',
            `app(\\JohnRivera7\\FilamentMia\\Settings\\Contracts\\SettingsRepository::class)` +
                `->put('preview', \\JohnRivera7\\FilamentMia\\Settings\\ThemeSettings::defaults()` +
                `->with(['login_layout' => '${layout}', 'login_tagline' => 'Client work, kept in one place.'])->toArray());`,
        ],
        { stdio: 'ignore' },
    )

    for (const mode of ['light', 'dark']) {
        const context = await browser.newContext({
            viewport: { width: 1440, height: 900 },
            colorScheme: mode,
        })

        await context.addInitScript(
            ([mode]) => localStorage.setItem('theme', mode),
            [mode],
        )

        const page = await context.newPage()
        await page.goto(`${BASE}/admin/login`, { waitUntil: 'networkidle' })

        // Put a validation error on screen, so its message is measured too.
        execFileSync('php', ['vendor/bin/testbench', 'tinker', '--execute', 'cache()->clear();'], {
            stdio: 'ignore',
        })
        await page.fill('[wire\\:model="data.email"]', 'john@mia.test')
        await page.fill('[wire\\:model="data.password"]', 'not-the-password')
        await page.locator('button[type=submit]').click()
        await page.waitForSelector('.fi-fo-field-wrp-error-message')
        await page.evaluate(() => document.fonts.ready)
        await page.waitForTimeout(600)

        for (const [name, selector] of TARGETS) {
            const target = page.locator(selector).first()

            if ((await target.count()) === 0 || !(await target.isVisible())) {
                continue
            }

            /*
             * Read back through a canvas rather than parsing the string.
             * Computed colours on this theme serialise as `oklch()`, because
             * that is how the palette is emitted, and a naive parse of those
             * three numbers as if they were RGB channels quietly produces
             * plausible-looking nonsense.
             */
            const colour = await target.evaluate((node) => {
                const canvas = document.createElement('canvas')
                canvas.width = canvas.height = 1

                const ctx = canvas.getContext('2d')
                ctx.fillStyle = getComputedStyle(node).color
                ctx.fillRect(0, 0, 1, 1)

                return [...ctx.getImageData(0, 0, 1, 1).data].slice(0, 3)
            })

            const box = await target.boundingBox()

            if (!box || box.width < 2 || box.height < 2) {
                continue
            }

            // Hide the glyphs without moving anything, then photograph the
            // box they were in. What is left is exactly what they sat on.
            await target.evaluate((node) => {
                node.style.color = 'transparent'
                node.style.textDecorationColor = 'transparent'
            })

            const file = join(
                dir,
                `${layout}-${mode}-${name.replace(/\s+/g, '-')}.png`,
            )

            await page.screenshot({ path: file, clip: box })

            await target.evaluate((node) => {
                node.style.removeProperty('color')
                node.style.removeProperty('text-decoration-color')
            })

            samples.push({ layout, mode, name, colour, file })
        }

        await context.close()
    }
}

await browser.close()

const manifest = join(dir, 'samples.json')
writeFileSync(manifest, JSON.stringify(samples))

try {
    execFileSync('python3', ['bin/contrast-login.py', manifest], {
        stdio: 'inherit',
    })
} catch {
    process.exitCode = 1
}
